tensorboard/uploader/proto/write_service.proto

```diff
@@ -119,6 +121,82 @@ message WriteScalarResponse {
   // This is empty on purpose.
 }
 
+// The blob writing procedure works as follows:
+//  * The caller instantiates a WriteBlob RPC, which is a bidirectional stream.
+//  * The caller sends a number of WriteBlobRequests, providing the blob data
+//    in an ordered sequence of chunks.
+//  * The request size should be less than 4 MiB (2^20 = 4,194,304), because
+//    that is the default request limit for gRPC.  Accounting for the other
+//    fields of the request, then, and leaving some padding for safety, the data
+//    chunk should be somewhat smaller-- perhaps simply 4e6 bytes.
+//  * The first of these requests must provide write_blob_first_request_data,
+//    and the subsequent ones must not.
+//  * The last of these requests must set finalize_object=True, and the prior
+//    ones must not.
+//  * For a blob <= 4MB, a single chunk suffices, so the single request is both
+//    the first and the last.
+//  * The caller receives a sequence of WriteBlobResponses in return.  Each
+//    WriteBlobResponse corresponds to one WriteBlobRequest in order, reporting
+//    the state of the write operation.
+//  * The caller may choose to verify each response before sending the next
+//    request, or it may continue sending requests optimistically.
```

Comment:
What does it mean to verify the response? If (e.g.) the response
contains a `size` other than the size that the client expects to have
seen, what can the client do?

Reply:
Right, it can check that the object_state and size are as expected, and raise an error if they're not, i.e. to avoid continuing to send data after the stream is already borked.  Perhaps we can rely on the RPC itself throwing errors in this case.  This is entangled with outstanding issues re whether the bidi-stream can be async (see below).

Reply:
I left the fields in the response because they are _sometimes_ useful, but the docstring now discourages validating every chunk, since that reduces throughput with little benefit.

Reply:
Further clarified RPC structure and docs around this.